firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexManager.java

```diff
@@ -75,14 +76,29 @@
   Collection<FieldIndex> getFieldIndexes();
 
   /**
-   * Returns an index that can be used to serve the provided target. Returns {@code null} if no
-   * index is configured.
+   * Returns an index that can be used to serve the provided target's conjunction filter.
+   *
+   * @param target The target for which we are looking for an index.
+   * @param andFilter The AND filter within the target's constraints for which we're looking for an
+   *     index. This can be null if the target has no filters.
+   * @return An index that can be used to serve the provided target's conjunction filter. Returns
+   *     {@code null} if no index is configured.
    */
   @Nullable
-  FieldIndex getFieldIndex(Target target);
+  FieldIndex getFieldIndex(Target target, @Nullable CompositeFilter andFilter);
```

Comment:
The reason this method exists is just so that the QueryEngine can quickly short-circuit Index-based execution if no index is defined. The callsite is:

```
FieldIndex i = getFIeldIndex();
if (i != null) {...}
```

It seems like this now adds some complication as we need to pass more state between `getFieldIndex` and `getDocumentsMatchingTarget`. We should probably revisit how this works to minimize these interactions (i.e. add a `hasMatchingIndex()` method or make `getDocumentsMatchingTarget()` nullable).

Reply:
Yes, you got it exactly right. I think I know how to simplify it now.